### PR TITLE
network: RouteTable= improvements

### PR DESCRIPTION
**Pull request by yuwata**
Follow-ups for c038ce4606f93d9e58147f87703125270fb744e2 (#18256).

### CODE REVIEW

**Comment by ssahani** [comment]
How about a test case .

File: src/network/networkd-route.c [hunk 1]
```diff
@@ -87,20 +87,94 @@ static const char * const route_table_table[] = {
...
+                str = strdup(s);
+                if (!str)
+                        return -ENOMEM;
 
-        return buf;
+                *ret = TAKE_PTR(str);
+                return 0;
+        }
+
+        if (m) {
```
Comment (ssahani) [inline src/network/networkd-route.c]: In which case manager will be `NULL` ?

File: src/network/networkd-route.c [hunk 2]
```diff
@@ -87,20 +87,94 @@ static const char * const route_table_table[] = {
...
+        if (table == 0)
+                return -EINVAL;
 
         s = route_table_to_string(table);
-        if (s)
-                strpcpy(&p, size, s);
-        else
-                strpcpyf(&p, size, "%d", table);
+        if (s) {
+                str = strdup(s);
```
Comment (ssahani) [inline src/network/networkd-route.c]: Here the route table name  (default/main/local)  with numbers can be added same as below.

File: src/network/networkd-route.c [hunk 3]
```diff
@@ -87,20 +87,94 @@ static const char * const route_table_table[] = {
...
+         * route_table_names_by_number, and then free the two hashmaps. */
+
+        while ((name = hashmap_steal_first(m->route_table_names_by_number))) {
+                hashmap_remove(m->route_table_numbers_by_name, name);
+                free(name);
+        }
+
+        m->route_table_names_by_number = hashmap_free(m->route_table_names_by_number);
+        m->route_table_numbers_by_name = hashmap_free(m->route_table_numbers_by_name);
 }
```
Comment (ssahani) [inline src/network/networkd-route.c]:
```
m->route_table_names_by_number = hashmap_free(m->route_table_names_by_number);
m->route_table_numbers_by_name = hashmap_free_free_key(m->route_table_numbers_by_name)
```

Then the above logic can be dropped.

Comment (yuwata) [reply src/network/networkd-route.c]: That's just for safety. An entry may be added in only one hashmap. Please see `config_parse_table_names()`. We have
```
if (r > 0 || k > 0)
    TAKE_PTR(name);
```
So, at least one of `hashmap_ensure_put()` may return 0.

Comment (yuwata) [reply src/network/networkd-route.c]: OK. By reconsidering your comments, I dropped the logic. Thanks.

**Comment by yuwata** [comment]
@ssahani Thank you for your comments. All your comments are addressed. Also, a test cases are added. PTAL.

**Comment by ssahani** [comment]
LGTM thanks

File: src/network/networkd-route.c [hunk 4]
```diff
@@ -87,20 +87,68 @@ static const char * const route_table_table[] = {
...
+int route_table_to_string_full(const Manager *m, uint32_t table, char **ret) {
+        _cleanup_free_ char *str = NULL;
         const char *s;
-        char *p = buf;
+
+        assert(m);
+        assert(ret);
+
+        if (table == 0)
+                return -EINVAL;
```
Comment (bluca) [inline src/network/networkd-route.c:131]: assert_return() maybe?

Comment (bluca) [reply src/network/networkd-route.c:131]: or is it expected to get 0 here from time to time?

Comment (yuwata) [reply src/network/networkd-route.c:131]: > or is it expected to get 0 here from time to time?

Should not, but might be. Currently, this is only used in debug logs. So, it should not trigger any assertions about `table`.

Note, usually `xxx_to_string()` does not trigger any assertion when an input is invalid, but simply returns NULL. Also similar for `xxx_to_string_alloc()`. So, I think it is better to simply return negative errno here.

Comment (bluca) [reply src/network/networkd-route.c:131]: Ok - note that I mentioned assert_return, ie: don't crash crash but logs an unexpected error

Comment (yuwata) [reply src/network/networkd-route.c:131]: Yeah, but an additional log in debug log is also not necessary.
See `log_route_debug()`, the function is void()ified.

Comment (bluca) [reply src/network/networkd-route.c:131]: ok, sounds good

File: src/network/networkd-route.c [hunk 5]
```diff
@@ -2401,47 +2398,65 @@ int config_parse_route_table_names(
...
-                return log_oom();
-        if (r == -EEXIST) {
-                log_syntax(unit, LOG_WARNING, filename, line, r,
-                           "Specified RouteTable= name and value pair conflicts with others, ignoring assignment: %s", rvalue);
-                return 0;
-        }
-        if (r > 0)
-                TAKE_PTR(name);
+                if (STR_IN_SET(name, "default", "main", "local")) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r,
```
Comment (poettering) [inline src/network/networkd-route.c]: r is not set here, 5th arg should be zero

File: src/network/networkd-route.c [hunk 6]
```diff
@@ -87,20 +87,68 @@ static const char * const route_table_table[] = {
         [RT_TABLE_LOCAL]   = "local",
 };
 
-DEFINE_STRING_TABLE_LOOKUP(route_table, int);
+DEFINE_PRIVATE_STRING_TABLE_LOOKUP(route_table, int);
 
-#define ROUTE_TABLE_STR_MAX CONST_MAX(DECIMAL_STR_MAX(int), STRLEN("default") + 1)
-static const char *format_route_table(int table, char *buf, size_t size) {
+int route_table_from_string_full(const Manager *m, const char *s, uint32_t *ret) {
```
Comment (poettering) [inline src/network/networkd-route.c]: i'd name these manager_get_route_table_by_string() btw...

the xyz_from_abc() and xyz_to_abc() functions are mostly stateless "pure" functions, i.e. that operate on completely static fixed tables. i.e. not dynamically registered stuff like this. 

but doesn't matter, this is just bikeshedding

**Comment by poettering** [comment]
also needs a rebase

**Comment by yuwata** [comment]
@poettering Thank you for your review. The above two points are addressed, and rebased on the current head. PTAL.

**Pull request merged**